Source/DafnyLanguageServer/Workspace/FinishedResolution.cs

```diff
@@ -0,0 +1,85 @@
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using Microsoft.Dafny.LanguageServer.Language;
+using Microsoft.Dafny.LanguageServer.Language.Symbols;
+using Microsoft.Dafny.LanguageServer.Workspace.Notifications;
+using Microsoft.Extensions.Logging;
+using OmniSharp.Extensions.LanguageServer.Protocol.Models;
+using Range = OmniSharp.Extensions.LanguageServer.Protocol.Models.Range;
+
+namespace Microsoft.Dafny.LanguageServer.Workspace;
+
+record FinishedResolution(
+  ImmutableDictionary<Uri, ImmutableList<Diagnostic>> Diagnostics,
+  Program ResolvedProgram,
+  SymbolTable? SymbolTable,
+  LegacySignatureAndCompletionTable LegacySignatureAndCompletionTable,
+  IReadOnlyDictionary<Uri, IReadOnlyList<Range>> GhostRanges,
+  IReadOnlyList<ICanVerify>? CanVerifies) : ICompilationEvent {
+  public IdeState UpdateState(DafnyOptions options, ILogger logger, IdeState previousState) {
+    var errors = Diagnostics.Values.SelectMany(x => x).
+      Where(d => d.Severity == DiagnosticSeverity.Error && d.Source != MessageSource.Compiler.ToString()).ToList();
+    var status = errors.Any() ? CompilationStatus.ResolutionFailed : CompilationStatus.ResolutionSucceeded;
+
+    var trees = previousState.VerificationTrees;
+    if (status == CompilationStatus.ResolutionSucceeded) {
+      foreach (var uri in trees.Keys) {
+        trees = trees.SetItem(uri,
+          GutterIconAndHoverVerificationDetailsManager.UpdateTree(options, ResolvedProgram,
+            previousState.VerificationTrees[uri]));
+      }
+    }
+
+    var verificationResults = CanVerifies == null
+      ? previousState.VerificationResults
+      : CanVerifies.GroupBy(l => l.NameToken.Uri).ToImmutableDictionary(k => k.Key,
+        k => k.GroupBy<ICanVerify, Range>(l => l.NameToken.GetLspRange()).ToImmutableDictionary(
+          l => l.Key,
+          l => MergeResults(l.Select(canVerify => MergeVerifiable(previousState, canVerify)))));
+    var signatureAndCompletionTable = LegacySignatureAndCompletionTable.Resolved ? LegacySignatureAndCompletionTable : previousState.SignatureAndCompletionTable;
+
+    return previousState with {
+      StaticDiagnostics = Diagnostics,
+      Status = status,
+      ResolvedProgram = ResolvedProgram,
+      SymbolTable = SymbolTable
+                    ?? previousState.SymbolTable, // TODO migration seems missing
```

Comment:
TODO: Can you please create an issue and link the issue to that todo, or fix it?